packages/workbox-background-sync/models/QueueStore.mjs

```diff
@@ -22,54 +26,179 @@ export class QueueStore {
    * Associates this instance with a Queue instance, so entries added can be
    * identified by their queue name.
    *
-   * @param {Queue} queue
-   *
+   * @param {string} queueName
    * @private
    */
-  constructor(queue) {
-    this._queue = queue;
-    this._db = new DBWrapper(DB_NAME, 1, {
-      onupgradeneeded: (evt) => evt.target.result
-          .createObjectStore(OBJECT_STORE_NAME, {autoIncrement: true})
-          .createIndex(INDEXED_PROP, INDEXED_PROP, {unique: false}),
+  constructor(queueName) {
+    this._queueName = queueName;
+    this._db = new DBWrapper(DB_NAME, DB_VERSION, {
+      onupgradeneeded: (evt) => this._upgradeDb(evt),
     });
   }
 
   /**
-   * Takes a StorableRequest instance, converts it to an object and adds it
-   * as an entry in the object store.
+   * Append an entry last in the queue.
    *
-   * @param {StorableRequest} storableRequest
+   * @param {Object} entry
+   * @param {Object} entry.requestData
+   * @param {number} [entry.timestamp]
+   * @param {Object} [entry.metadata]
+   */
+  async pushEntry(entry) {
+    if (process.env.NODE_ENV !== 'production') {
+      assert.isType(entry, 'object', {
+        moduleName: 'workbox-background-sync',
+        className: 'QueueStore',
+        funcName: 'pushEntry',
+        paramName: 'entry',
+      });
+      assert.isType(entry.requestData, 'object', {
+        moduleName: 'workbox-background-sync',
+        className: 'QueueStore',
+        funcName: 'pushEntry',
+        paramName: 'entry.requestData',
+      });
+    }
+
+    // Don't specify an ID since one is automatically generated.
+    delete entry.id;
+    entry.queueName = this._queueName;
+
+    await this._db.add(OBJECT_STORE_NAME, entry);
+  }
+
+  /**
+   * Preppend an entry first in the queue.
    *
-   * @private
+   * @param {Object} entry
+   * @param {Object} entry.requestData
+   * @param {number} [entry.timestamp]
+   * @param {Object} [entry.metadata]
    */
-  async addEntry(storableRequest) {
-    await this._db.add(OBJECT_STORE_NAME, {
-      queueName: this._queue.name,
-      storableRequest: storableRequest.toObject(),
-    });
+  async unshiftEntry(entry) {
```

Comment:
What do you think about exposing a [generator method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) as part of this class?

It would provide a nicer way of iterating over the items.

Reply:
As we discussed in chat, this probably won't work since we'd need it to be an async iterator, which isn't supported in all browsers yet. We'll revisit when that time comes.